src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -172,29 +180,86 @@ private Continuable createContinuable(CpsThread currentThread, CpsCallableInvoca
     }
 
     @Override
-    public synchronized Collection<StepExecution> getCurrentExecutions() {
-        if (thread==null)   return Collections.emptyList();
-
-        StepExecution s = thread.getStep();
-        if (s!=null)        return Collections.singleton(s);
-        else                return Collections.emptyList();
+    public Collection<StepExecution> getCurrentExecutions() {
+        CpsThread t;
+        synchronized (this) {
+            t = thread;
+            if (t == null) {
+                return Collections.emptySet();
+            }
+        }
+        final SettableFuture<Collection<StepExecution>> result = SettableFuture.create();
+        t.getExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
+            @Override public void onSuccess(CpsThreadGroup g) {
+                try {
+                    List<StepExecution> executions = new ArrayList<>();
+                    // cf. trick in CpsFlowExecution.getCurrentExecutions(true)
+                    Map<FlowHead, CpsThread> m = new LinkedHashMap<>();
+                    for (CpsThread t : g.threads.values()) {
+                        m.put(t.head, t);
+                    }
+                    for (CpsThread t : m.values()) {
+                        // TODO seems cumbersome to have to go through the flow graph to find out whether a head is a descendant of ours, yet FlowHead does not seem to retain a parent field
```

Comment:
So what I would really want _here_ is a way to get from a `FlowHead` to an `Iterable<CpsThread>` representing the parts of the call stack that interact with step executions, or conversely to go from a `CpsThread` to any descendants (`getNextInner` does not suffice). For example in

```groovy
timeout(…) {/* #1 */
  parallel a: {
    sleep 9999 /* #2 */
  }, b: {
    while (true) {/* #3 */}
  }
}
```

we would have a straightforward way of going from the `CpsBodyExecution` marked at point 1 to the `CpsThread`s representing points 2 and 3.